test/test_transform_collections_mixed.cpp

```diff
@@ -0,0 +1,1402 @@
+{
+    // Baseline: set 'any.A' to Dictionary and insert two integers in the dictionary
+    // {id: 1, any: {{"A": {{{"key1": 1}, {"key2": 2}}}}}}
+    TransformTestHarness h(test_context, TransformTestHarness::ClientOneBeforeTwo, [](Obj obj, ColKey col_any) {
+        obj.set_collection(col_any, CollectionType::Dictionary);
+        auto dict = obj.get_dictionary(col_any);
+        dict.insert_collection("A", CollectionType::Dictionary);
+        auto dict2 = dict.get_dictionary("A");
+        dict2->insert("key1", 1);
+        dict2->insert("key2", 2);
+    });
+
+    // Client 1 sets 'any.A' from Dictionary to List, inserts one integer in the list, clears the list, and inserts
+    // one more integer in the list
+    // {id: 1, any: {{"A": [2]}}}
+    h.transaction(h.client_1, [](Obj obj, ColKey col_any) {
+        auto dict = obj.get_dictionary(col_any);
+        dict.insert_collection("A", CollectionType::List);
+        auto list = dict.get_list("A");
+        list->add(1);
+        list->clear();
+        list->add(2);
+    });
+
+    // Client 2 sets 'any.A' from Dictionary to List and inserts one integer in the list
+    // {id: 1, any: {{"A": [4]}}}
+    h.transaction(h.client_2, [](Obj obj, ColKey col_any) {
+        auto dict = obj.get_dictionary(col_any);
+        dict.insert_collection("A", CollectionType::List);
+        auto list = dict.get_list("A");
+        list->add(4);
+    });
+
+    // Result: Client 1 wins - Clear wins against the insertion from Client 2
+    // {id: 1, any: {{"A": [2]}}}
+    h.check_merge_result([&](Obj obj, ColKey col_any) {
+        auto list = obj.get_list_ptr<Mixed>({col_any, "A"});
+        CHECK_EQUAL(list->size(), 1);
+        CHECK_EQUAL(list->get(0), 2);
+    });
+}
+
+TEST(Transform_ClearArrayVsCreateDictionary)
+{
+    // Baseline: set property 'any' to List
+    // {id: 1, any: []}
+    TransformTestHarness h(test_context, TransformTestHarness::ClientOneBeforeTwo, [](Obj obj, ColKey col_any) {
+        obj.set_collection(col_any, CollectionType::List);
+    });
+
+    // Client 1 inserts two integers in list at 'any', clears the list, and inserts one more integer
+    // {id: 1, any: [3]}
+    h.transaction(h.client_1, [](Obj obj, ColKey col_any) {
+        auto list = obj.get_list<Mixed>(col_any);
+        list.add(1);
+        list.add(2);
+        list.clear();
+        list.add(3);
+    });
+
+    // Client 2 sets property 'any' from List to Dictionary and inserts one integer in the dictionary
+    // {id: 1, any: {{"key1": 42}}}
+    h.transaction(h.client_2, [](Obj obj, ColKey col_any) {
+        obj.set_collection(col_any, CollectionType::Dictionary);
+        auto dict = obj.get_dictionary(col_any);
+        dict.insert("key1", 42);
+    });
+
+    // Result: Client 2 wins because changing the collection type is higher up in the path than updating the
+    // collection. Also, Clear wins against any updates on the collection from the other users
+    // {id: 1, any: {}}
```

Comment:
This is confusing to me: Since the `clear()` is being performed on the `any` column as a List, I wouldn't think it would still be applied to `any` now that it's a dict. Is it because the `clear()` instruction is technically being applied to `any` regardless of what type it currently is?
So the instructions are compared/merged like this?
* `add(1)` |  `set_collection` => 2nd wins for timestamp
* `add(2)` |  `insert("key1", 42)` => 2nd wins for timestamp and the fact that `any` is a dict
* `clear()` => clears the dictionary
* `add(3)` => discarded since `any` is not a list

Reply:
yeah, clear wins as long as the other type is **still** a collection. otherwise, it loses.

that's not how instructions are compared and transformed. every local instruction in the merging window is compared with every incoming instruction.

- `add(1)` | `set_collection` => 2nd wins if setting different collection type (add(1) is in the sub-path of set_collection)
- `add(2)` | `insert("key1", 42)` => they never get compared (it's add(2) | set_collection first, and set_collection wins)
- `clear()` => clears the dictionary
- `add(3)` => discarded because clear() wins (add(3) is in the sub-path of clear)